### PR TITLE
Cache requests for user's devices from federation

### PR DESCRIPTION
**Pull request by erikjohnston**
This should mitigate the issue where lots of different servers requests the same user's devices all at once.

### CODE REVIEW

File: synapse/storage/databases/main/end_to_end_keys.py [hunk 1]
```diff
@@ -114,6 +133,46 @@ async def get_e2e_device_keys_for_federation_query(
...
+            if not rows:
+                # No new rows, so cache is still valid.
+                return now_stream_id, cached_results
+
+            # There has, so let's invalidate the cache and run the query.
+            self._get_e2e_device_keys_for_federation_query_inner.invalidate((user_id,))
+
+        results = await self._get_e2e_device_keys_for_federation_query_inner(user_id)
+
+        return now_stream_id, results
```
Comment (squahtx) [inline synapse/storage/databases/main/end_to_end_keys.py:168]: I'm having trouble figuring out whether this is correct.

`cached_results` will come from a time when the device stream token was less than `now_stream_id`. But we only discard the cached result if there are rows with a stream ordering >= `now_stream_id`, so wouldn't we be overlooking changes between `cached_results`'s stream id and `now_stream_id`?

Comment (erikjohnston) [reply synapse/storage/databases/main/end_to_end_keys.py:168]: I think that case is fine, as for all changes less than `now_stream_id` we should have invalidated the cache?

Comment (squahtx) [reply synapse/storage/databases/main/end_to_end_keys.py:168]: I see, if `now_stream_id` can't advance without invalidating the cache then it looks okay.
We could probably add a comment to that effect.

**Review by squahtx — APPROVED** [review]
lgtm!

File: synapse/storage/databases/main/end_to_end_keys.py (hunk 1 above)
Comment (squahtx) [reply synapse/storage/databases/main/end_to_end_keys.py]: Yes, it makes enough sense.

**Pull request merged**